### DISCUSSION TITLE
Fabulous App Multilingual support

### DISCUSSION THREAD

**Discussion by edgarfgp** [fsprojects/Fabulous#901] [General]
I have been doing dome research about this topic  and and the main approach used is RESX files  as you can read here [link](https://devblogs.microsoft.com/xamarin/mastering-multilingual-in-xamarin-forms/).

1.  F# in general as far I know does not support resx files properly (We resx code behind is not generated . So we can not change the Culture programatically)
2.  Resx type provider does not work on NetStandard 2.1 and for instance will not support net5 or net6 with maui 

I have not still tried it . I guess the only alternative will be to create C# Library(Adding all the resx files and reference it from F#

I was wondering if you know different approach that we can use in Fabulous .

**Comment by TimLariviere** [comment]
If you really need to translate using RESX, yes, the best way is with a C# project.
There's also a RESX Type Provider, don't know what it's worth: https://fsprojects.github.io/FSharp.Configuration/ResXProvider.html

I initially wanted to add I18N to FabulousContacts and decided to externalize the strings in a module.
https://github.com/TimLariviere/FabulousContacts/blob/master/FabulousContacts/Strings.fs

So I guess you can go the same way by passing around the user language.
```fs
module Strings =
    let hello lang = match lang with EN -> "Hello" | FR -> "Bonjour"

View.Label(text = Strings.hello model.Language)
```
And trigger an update when user changes the language.


If you don't want to pass around the language, a stateful singleton is also an option
```fs
[<AbstractClass; Sealed>]
type Strings private () =
    static let strings = [
        "mainpage.hello", "Hello", "Bonjour"
    ]

    static let mutable _lang = EN

    static member SetLanguage(lang) = _lang <- lang

    static member Get(key) =
        let _, en, fr = strings |> List.find(fun (k,_,_) -> k = key)
        match _lang with
        | EN -> en
        | FR -> fr

View.Label(text = Strings.Get("mainpage.hello"))
```

**Reply by edgarfgp** [reply]
Thanks for the feedback. I tried https://fsprojects.github.io/FSharp.Configuration/ResXProvider.html . but unfortunately if does not work on net standard 2.1. I found a way to work with Resources.xx.resx and a little reflection  using the Thread.CurrentThread.CurrentCulture select the current translation

**Reply by edgarfgp** [reply]
AFAIK there is not a Global event to detect if the user has changed the device language setting . So I want be able to switch the culture while the app is one.

**Reply by delneg** [reply]
If you support iOS 13 or later, you can use system-provided language switching as per Apple guidelines:
https://developer.apple.com/news/?id=u2cfuj88
